File: Sample/Source/Windows/WindowCommand.py
# -*- encoding: utf-8 -*-

from __future__ import annotations
from typing import Any, Dict, Optional, Union, Tuple
from Engine import Pair, Image, IntRect, UI
from Engine.UI import ListView
from Engine.UI.FunctionalUI import FPlainText
from .Base import WindowSelectable


class WindowCommand(WindowSelectable):
    r"""\brief A simple selectable command list window.

    Provides a vertical list of command items with callbacks
    triggered on confirmation.
    """

    def __init__(
        self,
        rect: Union[IntRect, Tuple[Pair[int], Pair[int]]],
        commands: Dict[str, Dict[str, Any]] = {},
        rectWidth: Optional[int] = None,
        rectHeight: int = 32,
        windowSkin: Optional[Image] = None,
        repeated: bool = False,
        columns: int = 1,
    ) -> None:
        r"""\brief Construct a command window with the given commands.

        - \param rect The window rectangle.
        - \param commands Dictionary of command key to {text, callback}.
        - \param rectWidth Optional fixed width for the selection rectangle.
        - \param rectHeight Height of each command item.
        - \param windowSkin Optional window skin image.
        - \param repeated Whether the window skin is repeated.
        """
        super().__init__(rect, None, rectWidth, rectHeight, windowSkin, repeated)
        listView = ListView(self.content.getNoTranslationRect(), rectHeight, True, columns)
        if len(commands) > 0:
            for key, item in commands.items():
                child = FPlainText(UI.DefaultFont, item["text"], UI.DefaultFontSize)
                if "callback" in item:
                    child.addConfirmCallback(item["callback"])
                self._applyItem(child)
                listView.addChild(child)
        self.setListView(listView)

File: Sample/Scripts/Source/Windows/WindowCommand.lua
local WindowSelectable = require("Source.Windows.Base.WindowSelectable")
local WindowCommandController = require("Source.Windows.WindowCommand.Controller")

---@class Source.Windows.WindowCommand
local WindowCommand = {}

WindowCommand.controllerClass = WindowCommandController

function WindowCommand:init(rect, commands, rectWidth, rectHeight, windowSkin, repeated, columns, externalView)
    commands = commands or {}
    rectHeight = rectHeight or 32
    columns = columns or 1
    if externalView == nil then
        super(WindowCommand, self).init(rect, nil, rectWidth, rectHeight, windowSkin, repeated)
    else
        super(WindowCommand, self).init(rect, nil, rectWidth, rectHeight, windowSkin, repeated, nil, nil, true)
        if externalView.uiController ~= nil then
            externalView.uiController:attach()
            self:setScrollBox(externalView.uiController:getScrollBox())
            self:setListView(externalView.uiController:getListView())
        else
            self._window = externalView.windowFrame
            self.content = externalView.content
            self:setScrollBox(externalView.scrollBox)
            self:setListView(externalView.listView)
        end
    end
    self._commandController = self.controllerClass.new(self, self.content:getSize(), rectHeight, columns)
    if externalView == nil then
        self._commandController:attach(commands)
    else
        local listView = assert(self:getListView(), "External command ListView is unavailable")
        self._commandController:attachTo(listView, commands)
    end
end

function WindowCommand:refreshRows()
    self._commandController:refreshRows()
end

return class(WindowCommand, WindowSelectable)
